File: mobile/lib/models/product.dart
class Product {
  final int id;
  final String name;
  final String slug;
  final String? description;
  final double price;
  final double? compareAtPrice;
  final int? discountPercent;
  final String currency;
  final int? categoryId;
  final String? categoryName;
  final List<String> images;
  final int stockQty;
  final bool active;
  final bool inStock;
  final String createdAt;
  final double? avgRating;
  final int? reviewCount;

  const Product({
    required this.id,
    required this.name,
    required this.slug,
    this.description,
    required this.price,
    this.compareAtPrice,
    this.discountPercent,
    required this.currency,
    this.categoryId,
    this.categoryName,
    required this.images,
    required this.stockQty,
    required this.active,
    required this.inStock,
    required this.createdAt,
    this.avgRating,
    this.reviewCount,
  });

  String get primaryImage => images.isNotEmpty ? images.first : '';
  bool get isOnSale => compareAtPrice != null && compareAtPrice! > price;

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as int,
      name: json['name'] as String,
      slug: json['slug'] as String,
      description: json['description'] as String?,
      price: (json['price'] as num).toDouble(),
      compareAtPrice: (json['compareAtPrice'] as num?)?.toDouble(),
      discountPercent: json['discountPercent'] as int?,
      currency: json['currency'] as String? ?? 'INR',
      categoryId: json['categoryId'] as int?,
      categoryName: json['categoryName'] as String?,
      images: (json['images'] as List<dynamic>?)?.map((e) => e as String).toList() ?? [],
      stockQty: json['stockQty'] as int? ?? 0,
      active: json['active'] as bool? ?? true,
      inStock: json['inStock'] as bool? ?? true,
      createdAt: json['createdAt'] as String? ?? '',
      avgRating: (json['avgRating'] as num?)?.toDouble(),
      reviewCount: json['reviewCount'] as int?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'slug': slug,
      'description': description,
      'price': price,
      'compareAtPrice': compareAtPrice,
      'discountPercent': discountPercent,
      'currency': currency,
      'categoryId': categoryId,
      'categoryName': categoryName,
      'images': images,
      'stockQty': stockQty,
      'active': active,
      'inStock': inStock,
      'createdAt': createdAt,
      'avgRating': avgRating,
      'reviewCount': reviewCount,
    };
  }
}

File: backend/src/main/java/com/jaee/entity/Product.java
package com.jaee.entity;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Entity
@Table(name = "products")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false, unique = true)
    private String slug;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal price;

    /** Optional landed cost — for admin margin tracking only. */
    @Column(precision = 10, scale = 2)
    private BigDecimal baseCost;

    /** Billable shipping weight per unit (kg). */
    @Column(nullable = false, precision = 8, scale = 3)
    @Builder.Default
    private BigDecimal weightKg = new BigDecimal("0.5");

    @Column(precision = 10, scale = 2)
    private BigDecimal compareAtPrice;

    @Builder.Default
    private String currency = "INR";

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "product_categories",
            joinColumns = @JoinColumn(name = "product_id"),
            inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    @Builder.Default
    private Set<Category> categories = new HashSet<>();

    @ElementCollection
    @CollectionTable(name = "product_images", joinColumns = @JoinColumn(name = "product_id"))
    @Column(name = "image_url")
    @Builder.Default
    private List<String> images = new ArrayList<>();

    @ElementCollection
    @CollectionTable(name = "product_videos", joinColumns = @JoinColumn(name = "product_id"))
    @Column(name = "video_url")
    @Builder.Default
    private List<String> videos = new ArrayList<>();

    @ElementCollection
    @CollectionTable(name = "product_option_names", joinColumns = @JoinColumn(name = "product_id"))
    @Column(name = "option_name")
    @Builder.Default
    private List<String> options = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("sortOrder ASC, id ASC")
    @Builder.Default
    private List<ProductVariant> variants = new ArrayList<>();

    @Builder.Default
    private Integer stockQty = 0;

    @Builder.Default
    private Boolean active = true;

    @Builder.Default
    private LocalDateTime createdAt = LocalDateTime.now();

    // Review statistics (cached for performance)
    @Column(precision = 2, scale = 1)
    @Builder.Default
    private BigDecimal avgRating = BigDecimal.ZERO;

    @Builder.Default
    private Integer reviewCount = 0;

    public boolean isInStock() {
        return stockQty != null && stockQty > 0;
    }

    public void reduceStock(int quantity) {
        if (this.stockQty >= quantity) {
            this.stockQty -= quantity;
        } else {
            throw new IllegalStateException("Insufficient stock for product: " + name);
        }
    }
}
